src/data/program_configuration.js

```diff
@@ -367,18 +460,27 @@ export default class ProgramConfiguration {
 
 export class ProgramConfigurationSet<Layer: TypedStyleLayer> {
     programConfigurations: {[string]: ProgramConfiguration};
+    needsUpload: boolean;
 
     constructor(layoutAttributes: Array<StructArrayMember>, layers: $ReadOnlyArray<Layer>, zoom: number, filterProperties: (string) => boolean = () => true) {
         this.programConfigurations = {};
         for (const layer of layers) {
             this.programConfigurations[layer.id] = ProgramConfiguration.createDynamic(layer, zoom, filterProperties);
             this.programConfigurations[layer.id].layoutAttributes = layoutAttributes;
         }
+        this.needsUpload = false;
     }
 
-    populatePaintArrays(length: number, feature: Feature) {
+    populatePaintArrays(length: number, feature: Feature, index: number) {
         for (const key in this.programConfigurations) {
-            this.programConfigurations[key].populatePaintArrays(length, feature);
+            this.programConfigurations[key].populatePaintArrays(length, feature, index);
+        }
+        this.needsUpload = true;
+    }
+
+    updatePaintArrays(featureStates: FeatureStates, vtLayer: VectorTileLayer, layers: $ReadOnlyArray<TypedStyleLayer>) {
+        for (const layer of layers) {
+            this.needsUpload =  this.needsUpload || this.programConfigurations[layer.id].updatePaintArrays(featureStates, vtLayer, layer);
```

Comment:
If `this.needsUpload` is `true` before is gets called then the `||` will short-circuit and `this.programConfigurations[layer.id].updatePaintArrays(...)` will not get called. It should always get called, right?

Reply:
This is a great catch! I didn't run into the short-circuiting, but it is possible.